packages/scratch-vm/src/extensions/scratch3_spotify/index.ts

```diff
@@ -0,0 +1,492 @@
+            );
+        });
+    }
+
+    setupTimeouts() {
+        // events on each beat
+        this.beatTimeouts = [];
+        for (let i = 0; i < this.numBeats; i++) {
+            let t = window.setTimeout(
+                function (i) {
+                    this.beatFlag = true;
+                    this.currentBeatNum = i;
+                },
+                (this.trackTimingData.beats[i] - 0.1) * 1000,
+                i
+            );
+            this.beatTimeouts.push(t);
+        }
+
+        // events on each bar
+        this.barTimeouts = [];
+        for (let i = 0; i < this.trackTimingData.downbeats.length; i++) {
+            if (
+                this.trackTimingData.downbeats[i] <
+                this.trackTimingData.beats[this.numBeats - 1]
+            ) {
+                let t = window.setTimeout(() => {
+                    this.barFlag = true;
+                }, (this.trackTimingData.downbeats[i] - 0.1) * 1000);
+                this.barTimeouts.push(t);
+            }
+        }
+    }
+
+    // All example definitions below are syntactically equivalent,
+    // and which you use is just a matter of preference.
+    defineBlocks(): Spotify["BlockDefinitions"] {
+        // Example 1
+        //type DefineExampleCommand = DefineBlock<Blocks["exampleCommand"]>;
+
+        // Example 2
+        /*const exampleCommand: DefineExampleCommand = () => ({
+      type: BlockType.Command,
+      args: [ArgumentType.String, { type: ArgumentType.Number, defaultValue: 789 }],
+      text: (exampleString, exampleNumber) => `This is where the blocks display text goes, with arguments --> ${exampleString} and ${exampleNumber}`,
+      operation: (exampleString, exampleNumber, util) => {
+        alert(`This is a command! Here's what it received: ${exampleString} and ${exampleNumber}`); // Replace with what the block should do! 
+        console.log(util.stackFrame.isLoop); // just an example of using the BlockUtility
+      }
+    });*/
+
+        return {
+            // Example 1?
+            //exampleCommand,
+
+            // Example 3
+            /*exampleReporter: function (self: Spotify): Block<Blocks["exampleReporter"]> {
+        return {
+          type: BlockType.Reporter,
+          text: "This increments an internal field and then reports it's value",
+          operation: () => {
+            return ++self.exampleField;
+          }
+        }
+      },*/
+
+            // Example 4
+            // exampleHat: pickFromOptions,
+            searchAndPlay: searchAndPlay,
+            /*searchAndPlayWait: searchAndPlayWait,
+      playNextBeat: playNextBeat,
+      playBeat: playBeat,
+      playBeatAndWait: playBeatAndWait,*/
+            trackData: getTrackData,
+            /*currentBeat: getCurrentBeat,
+      everyBeat: everyBeat,
+      everyBar:everyBar,
+      stopMusic: stopMusic*/
+        };
+    }
+}
+
+// FEEDBACK is it redundant to have to define all of these blocks then make the function for them?
```

Comment:
Hm, could you clarify what you mean @randi-c-dubs ?

Reply:
Ignore this one, that was just me not understanding how typescript works